appserver/ejb/ejb-container/src/main/java/org/glassfish/ejb/deployment/annotation/handlers/AbstractEjbHandler.java

```diff
@@ -233,9 +233,9 @@ public HandlerProcessingResult processAnnotation(AnnotationInfo ainfo)
                                    ejbClass.getName() }));
                 return getDefaultFailedResult();
             }
-
-
-        } else {
+        }
+        // Create a new descriptor only if there are no entries in the ejb-jar with same class name. EJB 3.0 spec
+        else if(currentBundle.getEjbByClassName(ejbClass.getName()).length == 0) {
```

Comment:
>While handling the annotations, this elementName field is set to the name attribute of the corresponding EJB >annotation. In case name is not present, it is set to the default class name.

Have you tried this:  in the 'if' block, the block that @honghzzhang mentioned earlier,  if the name attribute of the annotation is not present, and getEjbByName(elementName) returns null (here elementName had been set to ejbClass.getSimpleName()),  then call getEjbByClassName(ejbClass.getName()) which should return the DD entry